zebra/kernel_netlink.c

```diff
@@ -781,19 +784,29 @@ static ssize_t netlink_send_msg(const struct nlsock *nl, void *buf,
  *
  * Returns -1 on error, 0 if read would block or the number of bytes received.
  */
-static int netlink_recv_msg(const struct nlsock *nl, struct msghdr msg,
-			    void *buf, size_t buflen)
+static int netlink_recv_msg(struct nlsock *nl, struct msghdr *msg)
 {
 	struct iovec iov;
 	int status;
 
-	iov.iov_base = buf;
-	iov.iov_len = buflen;
-	msg.msg_iov = &iov;
-	msg.msg_iovlen = 1;
+	iov.iov_base = nl->buf;
+	iov.iov_len = nl->buflen;
+	msg->msg_iov = &iov;
+	msg->msg_iovlen = 1;
 
 	do {
-		status = recvmsg(nl->sock, &msg, 0);
+		int bytes;
+
+		bytes = recv(nl->sock, NULL, 0, MSG_PEEK | MSG_TRUNC);
```

Comment:
This is a bit different from what I was expecting - I thought we'd look at the incoming message header and grow the buffer if necessary in order to accomodate large individual messages.

Will this return the number of bytes available to read in total, or does it know enough about the incoming messages' boundaries to know about just the next single netlink message? I know that we support/recommend large socket-level receive buffers, to accomodate a burst of messages at startup time, so I'm wondering if we'd end up with a very large nl->buf, much larger than we'd need for any one individual message?

Reply:
if we actually have the header it is much much too late. 

it returns the number of bytes available in the read in total.  The largest I've seen it grow to was 34k  I doubt it's going to be much larger

Reply:
but ... if we only need to read 32K, why do we need to make the socket-level receive buffers so large? I've assumed that we did that because we could have a large burst - like many MB - of data coming up via netlink, and we don't want the socket to discard those bytes before we have a chance to read them.

it seems like if we can PEEK, we could just peek at the header, and then decide how big a buffer we need in order to fetch the next waiting message. it seems like a stream socket, sort of, in that we need to take a look at the incoming header or framing in order to know how big the next message in the stream actually is, and then read until we've gotten that message.

Reply:
I think there is some conflation of different problems here.  The kernel can send chunks of data to us.  Each chunk can have X number of `struct nlmsghdr ` data structures with actual netlink data in that we must parse.  Additionally the kernel can pass data to us faster than we are at parsing it( say we parse a 16k chunk of data but it takes us 5 seconds to do so, in the mean time more data is being passed down the pipe from the kernel eventually filling it up )

Modifying the socket buffer sizes fixes the second problem ( or mitigates it really )..  When the kernel runs out of space to hold all the different messages, I believe we receive a ENOMEM and zebra exits, because we know the socket has dropped data( but what data? )

This code change is an attempt to handle the first situation, above.  We are telling the OS to hand us Y bytes, but in reality there is a read that should be Z bytes, and Y is < Z.    I've not seen this buffer go larger than 34k currently in my very limited testing.  I'll put a couple million routes in a test system and see what happens to the buffer sizes,  I suspect we will be ok and buffer sizes won't grow too much ( but who knows! )

Reply:
Donald's done the experiment that shows that we're getting back the next message size, not the entire waiting readable bytes, so this looks fine.

Reply:
As a note I just added 1 million routes to the linux kernel in addition to the interface changes in the bug that caused this issue.  There was only 1 buffer size increase and lots and lots of reads of 32k at a time of routes.

Reply:
Here is a possible optimization:
1. Before a read do `ioctl(socket, FIONREAD, &message_length)`
2. If current buffer size is less than needed `realloc`

It will save us from doing a message copy in case of truncated message.

> Donald's done the experiment that shows that we're getting back the next message size, not the entire waiting readable bytes, so this looks fine.

The netlink datagram may contain multiple `struct nlmsghdr` (as Donald mentioned), we could iterate over the available messages (check for `NLM_F_MULTI`) but I think getting the next datagram length is more efficient...

> I believe we receive a ENOMEM and zebra exits, because we know the socket has dropped data( but what data? )

That is true. We could attempt to fix that, but it'd probably require a lot of work. Here is an excerpt from the netlink(7) man page:

```
       Netlink is not a reliable protocol.  It tries its  best  to  deliver  a
       message  to  its  destination(s), but may drop messages when an out-of-
       memory condition or other error  occurs.   For  reliable  transfer  the
       sender  can request an acknowledgement from the receiver by setting the
       NLM_F_ACK flag.  An acknowledgement is an NLMSG_ERROR packet  with  the
       error  field  set to 0.  The application must generate acknowledgements
       for received messages itself.  The kernel tries to send an  NLMSG_ERROR
       message  for  every  failed  packet.  A user process should follow this
       convention too.
```

This looks optimal to avoid the out-of-memory issue and possibly fix memory / ENOMEM issues in smaller plataforms (e.g. smaller buffer required to receive all route installations).

Reply:
I am confused by your first comment, as that is exactly what is done.

Reply:
Sorry I read it wrong, you passed `NULL` for buffer so in theory it is equivalent to that `ioctl()` (I didn't think about it before).

Anyway the rest of the comment about ENOBUFS/ENOMEM still applies even though not directly relevant for this PR (buffer resizing).